cpp/include/graph.hpp

```diff
@@ -45,23 +67,16 @@ enum class DegreeDirection {
  * @tparam WT   Type of weight
  */
 template <typename VT, typename ET, typename WT>
-class GraphBase {
+class GraphViewBase {
```

Comment:
What's the purpose of this View class? (compare to creating a const reference to a non-view class?)

Reply:
I have just renamed GraphBase class to denote that it is a view. The class is already in place in cugraph.

Reply:
I am not sure I am following this. We have both GraphCOO, GraphCSR, GraphCSC AND GraphCOOView, GraphCSRView, GraphCSCView.

e.g.
https://github.com/rapidsai/cugraph/pull/799/files#diff-ba59ce02957abb53c35f1a1929fc0374R189
AND
https://github.com/rapidsai/cugraph/pull/799/files#diff-ba59ce02957abb53c35f1a1929fc0374R362

What are the purposes of non-view and view classes?

Reply:
The current model in the python code is that the python graph object owns a set of cudf columns that contain the COO representation of the graph, and possible a CSR and/or CSC version as well.

In the current code base, we have a GraphCOO, GraphCSR and GraphCSC that contain the device pointers to these arrays so that the C++ graph algorithms can access the data and execute the algorithms.

@kaatish is proposing renaming these classes to be a view of the device pointers (following the cudf design pattern, these three objects do not own the data, they simply have a view of this data to execute some algorithm).

For most algorithms this is sufficient.  However, there are a handful of algorithms that take a read-only graph as input (this view), but also return a graph as output.  This is the focus of what he's trying to grapple with here.

The non-view classes would be passed ownership of the data before being returned to the python layer.  The python layer would then claim ownership of the data and put it into a new python graph object.

Reply:
I did not understand the scope of your question before @seunghwak. So the main issue is that RMM_ALLOC and RMM_FREE will be deprecated.

In order to return a graph created at the C++, we currently allocate pointers and let the python layer take ownership. This won't be possible very soon due to changes in RMM ([issue](https://github.com/rapidsai/cugraph/issues/727)).

We tried to think of a few workarounds so that ALLOC_TRY usage would work transparently but it looks like it won't be possible. The only option according to me is to return a bunch of rmm::device_buffers in the graph class that python can take ownership of. That is why we need two graph classes. One to be a view, and the other to be an owning class that can enable moving of resources.

Reply:
Thanks for the clarification, and this makes sense.

One concern is that how we will handle more complex graph structures (e.g.Hornet style dynamic graphs) if python owns the memory; but we can revisit this once this actually happens.